Print the return values of seal API calls into the debug buffer (#1047)
hyperledger/solang: docs/code_gen_options.rst

This also means that, whenever the length of an array is accessed using '.length', it is replaced with a constant.

Note that this optimization does not cover every case. When an array is passed

> as a function argument, for instance, the length is unknown.

Reviewer: Please also change the first line and first paragraph of this file so it talks about code generation options, not just optimizations.

Debugging Options
=================

It may be desirable to have access to debug information regarding the contract execution. 

> How-ever, this might lead to less ideal code as well as much higher gas usage. Hence it is disabled by default,

Reviewer 2: ```suggestion
However, this might lead to more instructions as well as higher gas usage. Hence, it is disabled by default,
```
I think `less ideal` sounds awkward, so this suggestion is optional. Please, be sure to remove the dash from `how-ever` and place a comma after `hence`.

> but can be enabled using CLI flags. Just make sure to not use them in production deployments.

Reviewer 2: ```suggestion
but can be enabled using CLI flags. Just make sure not to use them in production deployments.
```

.. _log-api-calls:

Log runtime API call results
----------------------------

Runtime API calls are not necessarely guaranteed to succeed under all possible circumstances.
By design, the low revel results of these calls might be abstracted away in Solidity.
But for development purposes, it might be desirable to observe the low level result of such calls directly. 
The ``--log-api-calls`` flag will make the contract to print the low-level result of runtime calls (if there is any).

.. note::

>     For now, this will only have effect when used with the ``Substrate`` target.

Reviewer 2: ```suggestion
    For now, this only works for the ``Substrate`` target.
```
I think this sounds better.

This also means that, whenever the length of an array is accessed using '.length', it is replaced with a constant.

Note that this optimization does not cover every case. When an array is passed
as a function argument, for instance, the length is unknown.

Debugging Options
-----------------

It may be desirable to have access to debug information regarding the contract execution. 

> However, this might lead to more instructions as well as higher gas usage. Hence, it is disabled by default,

Reviewer 2: `might` is on line 236, not on 235.

but can be enabled using CLI flags. Just make sure not to use them in production deployments.

.. _log-api-return-codes:

Log runtime API call results
++++++++++++++++++++++++++++

Runtime API calls are not necessarely guaranteed to succeed under all possible circumstances.

> By design, the low level results of these calls might be abstracted away in Solidity.

Reviewer: You can't call any vm api in Solidity, they are always abstracted away

```suggestion
By design, the low level results of these calls are abstracted away in Solidity.
